src/string.cr

```diff
@@ -578,31 +578,79 @@ class String
+  # This method raises an exception if the string is not a valid float representation.
+  # If *whitespace* is `true`, the default, leading and trailing whitespace is ignored.
   #
   # ```
   # "123.45e1".to_f      # => 1234.5
   # "45.67 degrees".to_f # => 45.67
-  # "thx1138".to_f       # => 0.0
+  # "thx1138".to_f       # => ArgumentError
   # ```
-  def to_f
-    to_f64
+  def to_f(whitespace = true)
+    to_f64(whitespace: whitespace)
   end
 
-  # Returns the result of interpreting leading characters in this string as a floating point number (`Float32`).
-  # Extraneous characters past the end of a valid number are ignored. If there is not a valid number at the start of str,
-  # 0.0 is returned. This method never raises an exception.
+  # Returns the result of interpreting characters in this string as a floating point number (`Float64`).
+  # This method returns `nil` if the string is not a valid float representation.
+  # If *whitespace* is `true`, the default, leading and trailing whitespace is ignored.
+  #
+  # ```
+  # "123.45e1".to_f?      # => 1234.5
+  # "45.67 degrees".to_f? # => 45.67
+  # "thx1138".to_f?       # => nil
+  # ```
+  def to_f?(whitespace = true)
+    to_f64?(whitespace: whitespace)
+  end
+
+  # Returns the result of interpreting characters in this string as a floating point number (`Float32`).
+  # This method raises an exception if the string is not a valid float representation.
+  # If *whitespace* is `true`, the default, leading and trailing whitespace is ignored.
   #
   # See `#to_f`.
-  def to_f32
-    LibC.strtof self, nil
+  def to_f32(whitespace = true)
+    to_f32?(whitespace: whitespace) || raise ArgumentError.new("Invalid Float32: #{self}")
+  end
+
+  # Returns the result of interpreting characters in this string as a floating point number (`Float32`).
+  # This method returns `nil` if the string is not a valid float representation.
+  # If *whitespace* is `true`, the default, leading and trailing whitespace is ignored.
+  #
+  # See `#to_f`.
+  def to_f32?(whitespace = true)
+    to_f_impl(whitespace: whitespace) do
+      v = LibC.strtof self, out endptr
+      {v, endptr}
+    end
   end
 
   # Same as `#to_f`.
-  def to_f64
-    LibC.atof self
+  def to_f64(whitespace = true)
+    to_f64?(whitespace: whitespace) || raise ArgumentError.new("Invalid Float64: #{self}")
+  end
+
+  # Same as `#to_f?`.
+  def to_f64?(whitespace = true)
+    to_f_impl(whitespace: whitespace) do
+      v = LibC.strtod self, out endptr
+      {v, endptr}
+    end
+  end
+
+  private def to_f_impl(whitespace = true)
+    return unless whitespace || '0' <= self[0] <= '9' || self[0] == '-' || self[0] == '+'
+
+    v, endptr = yield
+    string_end = to_unsafe + bytesize
+
+    if whitespace
+      while endptr <= string_end && endptr.value.chr.whitespace?
+        endptr += 1
+      end
+    end
+
+    endptr >= string_end ? v : nil
```

Comment:
Maybe we can be more strict, since `endptr` must point to the last char, and we may add `strict` at the same time, since you already do check for it, by not allowing anything after except for whitespace:

```crystal
return v unless strict

if whitespace
  while endptr < string_end && endptr.value.chr.whitespace?
    endptr += 1
  end
end

v if endptr == string_end
```